File: muen2djs/graphics.js
/**
 * @namespace
 */

var graphics = Object.create({
    /**
     * @param {Color} color
     * @returns {typeof graphics}
     */
    clear: function (color) {
        __muenGraphicsClear(color);
        return this;
    },

    /**
     * @param {number} x
     * @param {number} y
     * @param {number} radius
     * @param {Color} color
     * @returns {typeof graphics}
     */
    circle: function (x, y, radius, color) {
        __muenGraphicsCircle(x, y, radius, color);
        return this;
    },

    /**
     * @param {number} x
     * @param {number} y
     * @param {number} width
     * @param {number} height
     * @param {Color} color
     * @returns {typeof graphics}
     */
    rectangle: function (x, y, width, height, color) {
        __muenGraphicsRectangle(x, y, width, height, color);
        return this;
    },

    /**
     * @param {Vector2} position
     * @param {Vector2} size
     * @param {Color} color
     * @returns {typeof graphics}
     */
    rectangleV: function (position, size, color) {
        __muenGraphicsRectangleV(position, size, color);
        return this;
    },

    /**
     * @param {Rectangle} rectangle
     * @param {Color} color
     * @returns {typeof graphics}
     */
    rectangleRec: function (rectangle, color) {
        __muenGraphicsRectangleRec(rectangle, color);
        return this;
    },

    /**
     * @param {Rectangle} rec
     * @param {Vector2} origin
     * @param {number} rotation
     * @param {Color} color
     * @returns {typeof graphics}
     */ // @ts-ignore
    rectanglePro: function (rec, origin, rotation, color) {
        __muenGraphicsRectanglePro(rec, origin, rotation, color);
        return this;
    },

    /**
     * @param {Camera} camera
     * @returns {typeof graphics}
     */ // @ts-ignore
    beginCameraMode: function (camera) {
        __muenGraphicsBeginCameraMode(camera);
        return this;
    },

    /**
     * @returns {typeof graphics}
     */ // @ts-ignore
    endCameraMode: function () {
        __muenGraphicsEndCameraMode();
        return this;
    },

    /**
     * @param {Texture} texture
     * @param {number} x
     * @param {number} y
     * @param {Color} tint
     * @returns {typeof graphics}
     */ // @ts-ignore
    texture: function (texture, x, y, tint) {
        __muenGraphicsTexture(texture, x, y, tint);
        return this;
    },

    /**
     * @param {Texture} texture
     * @param {Vector2} position
     * @param {Color} tint
     * @returns {typeof graphics}
     */ // @ts-ignore
    textureV: function (texture, position, tint) {
        __muenGraphicsTextureV(texture, position, tint);
        return this;
    },

    /**
     * @param {Texture} texture
     * @param {Vector2} position
     * @param {number} rotation
     * @param {number} scale
     * @param {Color} tint
     * @returns {typeof graphics}
     */ // @ts-ignore
    textureEx: function (texture, position, rotation, scale, tint) {
        __muenGraphicsTextureEx(texture, position, rotation, scale, tint);
        return this;
    },

    /**
     * @param {Texture} texture
     * @param {Rectangle} source
     * @param {Vector2} position
     * @param {Color} tint
     * @returns {typeof graphics}
     */ // @ts-ignore
    textureRec: function (texture, source, position, tint) {
        __muenGraphicsTextureRec(texture, source, position, tint);
        return this;
    },

    /**
     * @param {Texture} texture
     * @param {Rectangle} source
     * @param {Rectangle} dest
     * @param {Vector2} origin
     * @param {number} rotation
     * @param {Color} tint
     * @returns {typeof graphics}
     */ // @ts-ignore
    texturePro: function (texture, source, dest, origin, rotation, tint) {
        __muenGraphicsTexturePro(texture, source, dest, origin, rotation, tint);
        return this;
    },

    /**
     * @param {Texture} texture
     * @param {NPatch} nPatch
     * @param {Rectangle} dest
     * @param {Vector2} origin
     * @param {number} rotation
     * @param {Color} tint
     * @returns {typeof graphics}
     */ // @ts-ignore
    textureNPatch: function (texture, nPatch, dest, origin, rotation, tint) {
        __muenGraphicsTextureNPatch(texture, nPatch, dest, origin, rotation, tint);
        return this;
    },
});

module.exports = Object.freeze(graphics);

File: src/bindings/graphics.cpp
#include "graphics.hpp"

#include <raylib.h>

#include "bindings.hpp"
#include "bindings/types.hpp"
#include "js.hpp"

namespace muen::bindings::graphics {

auto define(js::State *j) -> void {
    define_global_function(
        j,
        [](js::State *j) -> void {
            const auto color = types::tocolor(j, 1);
            ClearBackground(color);
            js::pushundefined(j);
        },
        "__muenGraphicsClear",
        1
    );

    define_global_function(
        j,
        [](js::State *j) -> void {
            const auto x = js::tointeger(j, 1);
            const auto y = js::tointeger(j, 2);
            const auto radius = js::tofloat(j, 3);
            const auto color = types::tocolor(j, 4);
            DrawCircle(x, y, radius, color);
            js::pushundefined(j);
        },
        "__muenGraphicsCircle",
        4
    );

    define_global_function(
        j,
        [](js::State *j) -> void {
            const auto x = js::tointeger(j, 1);
            const auto y = js::tointeger(j, 2);
            const auto width = js::tointeger(j, 3);
            const auto height = js::tointeger(j, 4);
            const auto color = types::tocolor(j, 5);
            DrawRectangle(x, y, width, height, color);
            js::pushundefined(j);
        },
        "__muenGraphicsRectangle",
        5
    );

    define_global_function(
        j,
        [](js::State *j) -> void {
            const auto position = types::tovector2(j, 1);
            const auto size = types::tovector2(j, 2);
            const auto color = types::tocolor(j, 3);
            DrawRectangleV(position, size, color);
            js::pushundefined(j);
        },
        "__muenGraphicsRectangleV",
        3
    );

    define_global_function(
        j,
        [](js::State *j) -> void {
            const auto rec = types::torectangle(j, 1);
            const auto color = types::tocolor(j, 2);
            DrawRectangleRec(rec, color);
            js::pushundefined(j);
        },
        "__muenGraphicsRectangleRec",
        2
    );

    define_global_function(
        j,
        [](js::State *j) -> void {
            const auto rec = types::torectangle(j, 1);
            const auto origin = types::tovector2(j, 2);
            const auto rotation = static_cast<float>(js::tonumber(j, 3));
            const auto color = types::tocolor(j, 4);
            DrawRectangleRec(rec, color);
            js::pushundefined(j);
        },
        "__muenGraphicsRectanglePro",
        4
    );

    define_global_function(
        j,
        [](js::State *j) -> void {
            const auto camera = types::tocamera(j, 1);
            BeginMode2D(camera);
            js::pushundefined(j);
        },
        "__muenGraphicsBeginCameraMode",
        1
    );

    define_global_function(
        j,
        [](js::State *j) -> void {
            EndMode2D();
            js::pushundefined(j);
        },
        "__muenGraphicsEndCameraMode"
    );

    define_global_function(
        j,
        [](js::State *j) -> void {
            const auto texture = types::totexture(j, 1);
            const auto x = js::tointeger(j, 2);
            const auto y = js::tointeger(j, 3);
            const auto tint = types::tocolor(j, 4);
            DrawTexture(texture, x, y, tint);
            js::pushundefined(j);
        },
        "__muenGraphicsTexture",
        5
    );

    define_global_function(
        j,
        [](js::State *j) -> void {
            const auto texture = types::totexture(j, 1);
            const auto position = types::tovector2(j, 2);
            const auto tint = types::tocolor(j, 3);
            DrawTextureV(texture, position, tint);
            js::pushundefined(j);
        },
        "__muenGraphicsTextureV",
        4
    );

    define_global_function(
        j,
        [](js::State *j) -> void {
            const auto texture = types::totexture(j, 1);
            const auto position = types::tovector2(j, 2);
            const auto rotation = js::tofloat(j, 3);
            const auto scale = js::tofloat(j, 4);
            const auto tint = types::tocolor(j, 5);
            DrawTextureEx(texture, position, rotation, scale, tint);
            js::pushundefined(j);
        },
        "__muenGraphicsTextureEx",
        6
    );

    define_global_function(
        j,
        [](js::State *j) -> void {
            const auto texture = types::totexture(j, 1);
            const auto source = types::torectangle(j, 2);
            const auto position = types::tovector2(j, 3);
            const auto tint = types::tocolor(j, 4);
            DrawTextureRec(texture, source, position, tint);
            js::pushundefined(j);
        },
        "__muenGraphicsTextureRec",
        5
    );

    define_global_function(
        j,
        [](js::State *j) -> void {
            const auto texture = types::totexture(j, 1);
            const auto source = types::torectangle(j, 2);
            const auto dest = types::torectangle(j, 3);
            const auto origin = types::tovector2(j, 4);
            const auto rotation = js::tofloat(j, 5);
            const auto tint = types::tocolor(j, 6);
            DrawTexturePro(texture, source, dest, origin, rotation, tint);
            js::pushundefined(j);
        },
        "__muenGraphicsTexturePro",
        7
    );

    define_global_function(
        j,
        [](js::State *j) -> void {
            const auto texture = types::totexture(j, 1);
            const auto nPatch = types::tonpatch(j, 2);
            const auto dest = types::torectangle(j, 3);
            const auto origin = types::tovector2(j, 4);
            const auto rotation = js::tofloat(j, 5);
            const auto tint = types::tocolor(j, 6);
            DrawTextureNPatch(texture, nPatch, dest, origin, rotation, tint);
            js::pushundefined(j);
        },
        "__muenGraphicsTextureNPatch",
        7
    );
}

} // namespace muen::bindings::graphics
